### DISCUSSION TITLE
Stroke based fonts

### DISCUSSION THREAD

**Discussion by bensapirstein** [coldtype/coldtype#132] [Q&A]
I would like to create stroke-based letterforms and animate them similar to outline-based fonts. Is this feature currently supported by Coldtype?

**Comment by stenson** [comment]
Definitely possible with Coldtype — here’s an example using a stroke font I have on my computer. Tough part is generating a font with unclosed characters, though, before you even get to the Coldtype part. Here I'm using a UFO file directly (internally compiled by Coldtype to a ttf w/o overlap removal). How are you making/generating the stroke font?

```python
from coldtype import *

@animation((1080, 320), bg=1)
def scratch(f):
    stroke_font = Font.Find("Hershey-DuplexPrincipalItalic.ufo")
    return (StSt("Stroke\nFont", stroke_font, 130)
        .lead(20)
        .xalign(f.a.r)
        .align(f.a.r)
        .fssw(-1, hsl(0.7), 1))
```

![scratch_0000](https://user-images.githubusercontent.com/89770/235255835-a2983587-9f5b-49d4-b774-e8cc3b791d5c.png)

**Reply by bensapirstein** [reply]
Thanks :)

I am currently using Adobe Illustrator to draw stroke-based letterforms and the blending tool to create interpolations. Exporting the frames is quite time-consuming and the animation curve is linear, so imagine how thrilled I was to discover Coldtype. While I am not fixed on any specific font creation routine, I am looking for a way to create variable stroke-based fonts. I plan to experiment with [Fontra](https://github.com/googlefonts/fontra) once it has good enough support for variable fonts, but I may start with designing some basic glyphs under the UFO format to create a proof of concept. My ultimate goal is to create interpolations between Hebrew and Arabic letterforms, which can only be achieved with stroke-based representations. I reached out to you on Twitter with more details about my project. Here's an example of the interpolation frames I'm working on, this was first designed in AI then exported to SVG and loaded to python to create the animation:

![Alf_1](https://user-images.githubusercontent.com/14977606/235292878-44555472-d41e-479c-864f-ce323b0ade57.gif)


(Alef in Hebrew and in Arabic)

**Reply by stenson** [reply]
Very cool project!

That makes sense re: Fontra. Coldtype itself can do some interpolation w/o having to build an actual variable font, as well, if you have two ufo sources, which either RoboFont or Glyphs can make.

Here's a quick example of that kind of interpolation

```python
from coldtype import *
from coldtype.fx.skia import phototype

ufo1 = Font.Cacheable(__sibling__("stroke1.ufo"))
ufo2 = Font.Cacheable(__sibling__("stroke2.ufo"))

@animation(tl=60, bg=bw(1))
def broad_nib(f):
    return (StSt("a", ufo1, 750).pen()
        .interpolate(f.e("eeio"), StSt("a", ufo2, 750).pen())
        .outline(0.1)
        .castshadow(33, 150)
        .align(f.a.r)
        .f(1)
        .ch(phototype(f.a.r, blur=3, cut=90, cutw=30, fill=bw(0))))
```

https://user-images.githubusercontent.com/89770/235392075-f4c598b1-158e-4af1-86ec-34e3cf0c5560.mp4

**Reply by bensapirstein** [reply]
I did it! :)

I took existing GLIF files, [converted them to svg](https://github.com/ctrlcctrlv/glif2svg/blob/main/bin/glif2svg), traced the strokes using AI, saved as SVG, and [converted them back to glifs](https://github.com/googlefonts/fonttools/blob/515270d874b30611f2148d4cc170f6236d34db9a/Snippets/svg2glif.py) I then loaded the UFOs and interpolated them, as you suggested. As you can see, I added a vertical flip because the process flipped the design. I believe this has to do with the conversion from SVG.

However, one issue I encountered was creating a proper stroke-based variable font. When I created the TTF font file using [fontmake](https://github.com/googlefonts/fontmake) it closed the open outlines, which ruined the stroke representation. I think it's time to consider proper font editing software like RoboFont.

```python
from coldtype import *
from coldtype.fx.skia import phototype

fnt_ar = Font.Find("Wasim-Ar")
fnt_hb = Font.Find("Wasim-Heb")

@animation(tl=120, bg=bw(1))
def scratch(f):
    return (StSt("א", fnt_ar, 1000).pen()
            .interpolate(f.e("eeio"), StSt("א", fnt_hb, 1000).pen())
            .explode()
            .scale(1,-1)
            .outline(0.1)
            .castshadow(40, 60 + 40 * f.e("eeio"))
            .align(f.a.r)
            .f(1)
            .ch(phototype(f.a.r, blur=3, cut=90, cutw=30, fill=bw(0)))
            )
```

https://user-images.githubusercontent.com/14977606/235948730-6a85478f-4423-4d17-a79f-227be7c082f3.mp4

**Reply by stenson** [reply]
So cool! Glad you got it to work. Completely unimportant but you can express the `f.e` 60-100 range as `f.e("eeio", rng=(60, 100))`

As for `fontmake`, what if you pass `--keep-overlaps` as an option on the command line?

**Comment by bensapirstein** [comment]
Thank you very much! 

I have one more question: does Coldtype support animating the drawing of strokes?
For reference, I created this animation frame by frame using Adobe Illustrator:
![Lisan](https://user-images.githubusercontent.com/14977606/235439654-177c3236-0b2f-4647-9bfb-a1832a1868ed.gif)

**Reply by stenson** [reply]
It does!

```python
from coldtype import *

ufo1 = Font.Cacheable(__sibling__("stroke2.ufo"))

@animation(tl=60, bg=bw(1))
def broad_nib(f):
    return (StSt("a", ufo1, 750).pen()
        .explode()
        .align(f.a.r)
        .map(lambda p: p
            .subsegment(0, f.e("ceio"))
            .outline(0.1)
            .castshadow(33, 100))
        .f(hsl(0.65, 1, 0.6, 0.6)))
```

https://user-images.githubusercontent.com/89770/235479182-6c9e5f09-cacf-4fcb-93dd-992185140e94.mp4
